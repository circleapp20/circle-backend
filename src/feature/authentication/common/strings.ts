import { VERIFICATION_CODE_EMAIL_TEMPLATE } from 'feature/authentication/common/templates';

export const getVerificationEmailText = (code: string, user = 'User') => {
	const template = VERIFICATION_CODE_EMAIL_TEMPLATE;
	return template.replace('{verificationCode}', code).replace('{user}', user);
};
