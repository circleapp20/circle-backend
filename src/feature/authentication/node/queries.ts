import { Users } from 'shared/common/schema/users';
import { EntityManager } from 'typeorm';

export const countMatchingIdAndCodeQuery = (
	manager: EntityManager,
	id: string,
	verificationCode: string
) => {
	return manager
		.createQueryBuilder(Users, 'u')
		.where('u.id = :id', { id })
		.andWhere('u.verificationCode = :code', { code: verificationCode })
		.getCount();
};

export const updateUserVerificationCodeQuery = (
	manager: EntityManager,
	values: { id: string; verificationCode: string }
) => {
	const { id, verificationCode } = values;
	return manager
		.createQueryBuilder(Users, 'u')
		.update()
		.set({ verificationCode })
		.where('u.id = :id', { id })
		.execute();
};
