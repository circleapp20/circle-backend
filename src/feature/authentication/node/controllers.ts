import { Response } from 'express';
import {
	checkUserVerificationCode,
	createUserProfileWithDefaultValues,
	getUserAccountWithCredentials,
	getUserProfileById,
	verifyUserLoginCredentials
} from 'feature/authentication/node/dataService';
import { sendVerificationCodeByMedia } from 'feature/authentication/node/messaging';
import { getResponseData } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { IAuthUser, IRequest } from 'shared/types';

export const verifyUserCredentials = async (req: IRequest, res: Response) => {
	const user = await createUserProfileWithDefaultValues(req.body.data);

	sendVerificationCodeByMedia({
		media: user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserVerificationCode = async (req: IRequest, res: Response) => {
	const { id }: IAuthUser = req.user;
	const data = Object.assign({}, req.body.data, { id });
	const status = await checkUserVerificationCode(data);
	const responseData = getResponseData(status);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserLogin = async (req: IRequest, res: Response) => {
	const user = await verifyUserLoginCredentials(req.body.data);
	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const resendUserVerificationCode = async (req: IRequest, res: Response) => {
	const { media }: any = req.query;
	const user = await getUserProfileById(req.user.id);

	sendVerificationCodeByMedia({
		media: media ? media : user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(true);
	res.status(Constants.status.CREATED).json(responseData);
};

/**
 * the controller verifies the user's credentials in the database
 * and sends the verification code to the user by email or password
 * should the user have both email and password, a response of message
 * is return to the user to request which media to send the verification
 * code to either email or phone number
 */
export const verifyUserCredentialsForPasswordReset = async (req: IRequest, res: Response) => {
	const profile = await getUserAccountWithCredentials(req.body.data);

	const { email, verificationCode, phoneNumber } = profile;
	let message = 'Verification code cannot be sent. User has both email and phone number';

	sendVerificationCodeByMedia({
		media: phoneNumber ? 'phoneNumber' : 'email',
		verificationCode,
		email: email && !phoneNumber ? email : '',
		phoneNumber: !email && phoneNumber ? phoneNumber : ''
	});

	if (email && !phoneNumber) {
		message = "Verification code sent to user's email";
	} else if (!email && phoneNumber) {
		message = "Verification code sent to user's phone number";
	}

	const responseData = getResponseData({ user: profile, message });
	res.status(Constants.status.CREATED).json(responseData);
};
