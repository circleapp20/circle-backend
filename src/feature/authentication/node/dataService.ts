import bcryptjs from 'bcryptjs';
import {
	countMatchingIdAndCodeQuery,
	updateUserVerificationCodeQuery
} from 'feature/authentication/node/queries';
import { getBadRequestError } from 'shared/common/errors';
import { generateCodeFromNumber } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { runInsertQuery, runInTransaction, runQuery } from 'shared/node/database';
import {
	addUserProfileQuery,
	getUserByCredentialsQuery,
	getUserByIdQuery
} from 'shared/node/queries';
import { getSignedAuthToken } from 'shared/node/validation';
import { IAddUserProfile } from 'shared/types';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode,
			roles: [Constants.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(addUserProfileQuery, [profile], manager);

		return runQuery(getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));

	// add token to user profile
	const token = getSignedAuthToken({ id: userProfile!.id, roles: [Constants.privileges.USER] });
	const profile = Object.assign({}, userProfile, { token });

	return profile;
};

export const checkUserVerificationCode = async (data: { id: string; verificationCode: string }) => {
	const { id, verificationCode } = data;

	if (!id || !verificationCode) {
		throw getBadRequestError('user id and verification code are required');
	}

	const count = await runQuery(countMatchingIdAndCodeQuery, [id, verificationCode]);
	if (!count) throw getBadRequestError('invalid verification code');

	return true;
};

export const verifyUserLoginCredentials = async (data: {
	email?: string;
	phoneNumber?: string;
	username?: string;
	password: string;
}) => {
	const { phoneNumber, password, email, username } = data;
	if (!phoneNumber && !email && !username) {
		throw getBadRequestError('phoneNumber, email or username is required');
	}
	if (!password) throw getBadRequestError('password is required');

	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (!bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Invalid user account');
	}

	const { password: hashedPassword, ...other } = user;

	// create user token
	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token });

	return profile;
};

export const getUserProfileById = async (id: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');
	return user;
};

/**
 * the service function validates the user's credentials in the
 * database to check if the user exists
 * if the user does exists, a new verification code is generated
 * and stored in the database
 */
export const getUserAccountWithCredentials = async (args: {
	email?: string;
	phoneNumber?: string;
	username?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [args]);
	if (!user) throw getBadRequestError('Invalid user account');

	// generate new verification code
	const verificationCode = generateCodeFromNumber();

	// update user profile
	await runQuery(updateUserVerificationCodeQuery, [{ id: user.id, verificationCode }]);

	const { password, ...other } = user;

	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token, verificationCode });

	return profile;
};
