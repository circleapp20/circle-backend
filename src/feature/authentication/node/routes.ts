import * as schema from 'feature/authentication/common/dataSchema';
import * as controller from 'feature/authentication/node/controllers';
import { Constants } from 'shared/constants';
import { IApiRoute } from 'shared/types';

export const authRoutes: IApiRoute[] = [
	{
		path: '/auth/sign-up',
		method: 'post',
		controller: controller.verifyUserCredentials,
		schema: schema.VerifyUserCredentialsSchema
	},
	{
		path: '/auth/code',
		method: 'post',
		controller: controller.verifyUserVerificationCode,
		privileges: [Constants.privileges.SUPER_ADMIN, Constants.privileges.USER],
		schema: schema.VerifyVerificationCodeSchema
	},
	{
		path: '/auth/sign-in',
		method: 'post',
		controller: controller.verifyUserLogin,
		schema: schema.VerifyUserLoginSchema
	},
	{
		path: '/auth/code/resend',
		method: 'post',
		controller: controller.resendUserVerificationCode,
		privileges: [Constants.privileges.USER]
	},
	{
		path: '/auth/credentials',
		method: 'post',
		controller: controller.verifyUserCredentialsForPasswordReset
	}
];
