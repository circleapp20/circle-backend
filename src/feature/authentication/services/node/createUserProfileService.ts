import { getBadRequestError } from 'base/common/errors';
import { generateCodeFromNumber } from 'base/common/utilities';
import { Constants } from 'base/constants';
import * as sharedQueries from 'base/node/queries';
import { getSignedAuthToken } from 'base/node/validation';
import { IAddUserProfile } from 'base/types';
import { runInsertQuery, runInTransaction, runQuery } from 'core/node/database/queryRunners';
import { decryptData, encryptData } from 'core/node/encryption';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();
		const encryptedVerificationCode = encryptData({ text: verificationCode });

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode: encryptedVerificationCode,
			roles: [Constants.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(sharedQueries.addUserProfileQuery, [profile], manager);

		return runQuery(sharedQueries.getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(sharedQueries.getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));
	const verificationCode = decryptData({ encryptedText: userProfile!.verificationCode });

	// add token to user profile
	const token = getSignedAuthToken({ id: userProfile!.id, roles: [Constants.privileges.USER] });
	const profile = Object.assign({}, userProfile, { token, verificationCode });

	return profile;
};
