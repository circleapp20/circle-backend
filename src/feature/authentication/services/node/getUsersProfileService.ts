import { getBadRequestError } from 'base/errors/node/badRequestError';
import { runQuery } from 'core/database/queryRunners';
import { decryptData } from 'core/encryption/node/encryption';
import { getUserByIdQuery } from 'core/queries/userQueries';

export const getUserProfileById = async (id: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');
	const { password, verificationCode, ...other } = user;
	const decryptedVerificationCode = decryptData({ encryptedText: verificationCode });
	return Object.assign({}, other, { verificationCode: decryptedVerificationCode });
};
