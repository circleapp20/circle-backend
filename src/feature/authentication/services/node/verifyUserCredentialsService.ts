import { getBadRequestError } from 'base/errors/node/badRequestError';
import { getSignedAuthToken } from 'base/server/validation';
import { generateCodeFromNumber } from 'base/utils/node/codeGenerator';
import bcryptjs from 'bcryptjs';
import { runQuery } from 'core/database/queryRunners';
import { decryptData, encryptData } from 'core/encryption/node/encryption';
import { getUserByCredentialsQuery, getUserByIdQuery } from 'core/queries/userQueries';
import { updateUserVerificationCodeQuery } from 'feature/authentication/queries/updateUserVerificationCodeQuery';

export const verifyUserLoginCredentials = async (data: {
	email?: string;
	phoneNumber?: string;
	username?: string;
	password: string;
}) => {
	const { phoneNumber, password, email, username } = data;
	if (!phoneNumber && !email && !username) {
		throw getBadRequestError('phoneNumber, email or username is required');
	}
	if (!password) throw getBadRequestError('password is required');

	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (!bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Invalid user account');
	}

	const { password: hashedPassword, verificationCode, ...other } = user;

	// create user token
	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token });

	return profile;
};

/**
 * the service function validates the user's credentials in the
 * database to check if the user exists
 * if the user does exists, a new verification code is generated
 * and stored in the database
 */
export const getUserAccountWithCredentials = async (args: {
	email?: string;
	phoneNumber?: string;
	username?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [args]);
	if (!user) throw getBadRequestError('Invalid user account');

	// generate new verification code
	const verificationCode = generateCodeFromNumber();
	const encryptedVerificationCode = encryptData({
		text: verificationCode
	});

	// update user profile
	await runQuery(updateUserVerificationCodeQuery, [
		{ id: user.id, verificationCode: encryptedVerificationCode }
	]);

	const { password, ...other } = user;

	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token, verificationCode });

	return profile;
};

export const checkUserVerificationCode = async (data: { id: string; verificationCode: string }) => {
	const { id, verificationCode } = data;

	if (!id || !verificationCode) {
		throw getBadRequestError('user id and verification code are required');
	}

	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');

	const decryptedVerificationCode = decryptData({ encryptedText: user.verificationCode });
	if (verificationCode !== decryptedVerificationCode) {
		throw getBadRequestError('Invalid verification code');
	}

	return true;
};
