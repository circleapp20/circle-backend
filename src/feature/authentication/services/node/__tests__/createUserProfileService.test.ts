import * as utils from 'base/common/utilities';
import { Constants } from 'base/constants';
import * as sharedQueries from 'base/node/queries';
import { entityManager } from 'base/testUtils/node/entityManager';
import * as encryption from 'core/node/encryption';
import faker from 'faker';
import { createUserFixture } from 'fixtures/users';
import {
	addUserTransaction,
	createUserProfileWithDefaultValues
} from '../createUserProfileService';

jest.mock('core/node/database/queryRunners');
jest.mock('base/common/schema/users');
jest.mock('base/common/schema/fellows');
jest.mock('bcryptjs', () => ({
	hashSync: jest.fn().mockReturnValue('$$20yy39nv93n932n92093nf92'),
	compareSync: jest.fn()
}));

describe('#addUserTransaction', () => {
	const addUserQuerySpy = jest.spyOn(sharedQueries, 'addUserProfileQuery');
	let createUserTransaction: (manager: any) => Promise<any>;

	beforeEach(() => {
		entityManager.execute.mockReturnValueOnce({
			generateMaps: [{ id: 'x7i9-3l-n3k4-3i8bi2' }]
		});
		createUserTransaction = addUserTransaction('test@test.com');
	});

	test('should create user with email', async () => {
		await createUserTransaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalled();
	});

	test('should create user with phoneNumber', async () => {
		const transaction = addUserTransaction('', '+2339874563');
		await transaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalled();
	});

	test('should generate random code for verification', async () => {
		const spy = jest.spyOn(utils, 'generateCodeFromNumber');
		await createUserTransaction(entityManager);
		expect(spy).toHaveBeenCalledWith();
	});

	test('should encrypt verification code', async () => {
		const spy = jest.spyOn(encryption, 'encryptData');
		await createUserTransaction(entityManager);
		expect(spy).toHaveBeenCalledWith(expect.objectContaining({ text: expect.any(String) }));
	});

	test('should add the user with the default profile', async () => {
		await createUserTransaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalledWith(
			entityManager,
			expect.objectContaining({
				username: '',
				password: '',
				email: expect.stringMatching('test@test.com'),
				phoneNumber: '',
				roles: expect.arrayContaining([Constants.privileges.USER]),
				verificationCode: expect.any(String)
			})
		);
	});

	test('should return the added user', async () => {
		const getUserQuerySpy = jest.spyOn(sharedQueries, 'getUserByIdQuery');
		await createUserTransaction(entityManager);
		expect(getUserQuerySpy).toHaveBeenCalledWith(entityManager, 'x7i9-3l-n3k4-3i8bi2');
	});

	test('should return the new user profile', async () => {
		entityManager.getOne.mockReturnValueOnce({
			id: 'x7i9-3l-n3k4-3i8bi2',
			verificationCode: '$$20yy39nv93n932n92093nf92',
			phoneNumber: '',
			email: 'test@test.com',
			roles: ['user']
		});
		const results = await createUserTransaction(entityManager);
		expect(results).toEqual(
			expect.objectContaining({
				id: expect.any(String),
				email: expect.stringMatching('test@test.com'),
				phoneNumber: '',
				roles: expect.arrayContaining([Constants.privileges.USER]),
				verificationCode: expect.any(String)
			})
		);
	});
});

describe('#createUserProfileWithDefaultValues', () => {
	const verificationCode = encryption.encryptData({ text: faker.random.alphaNumeric() });
	const profile = { ...createUserFixture(), verificationCode };

	test('should add user if email does not exists', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		const user = await createUserProfileWithDefaultValues({ email: profile.email });
		expect(user).toEqual(expect.objectContaining({ token: expect.any(String) }));
	});

	test('should create new user if phoneNumber does not exists', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		const user = await createUserProfileWithDefaultValues({
			phoneNumber: profile.phoneNumber
		});
		expect(user).toEqual(expect.objectContaining({ token: expect.any(String) }));
	});

	test('should throw an error if email already exists', () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		expect(
			createUserProfileWithDefaultValues({ email: 'test@test.com' })
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw bad request error if phone number already exists', () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		expect(
			createUserProfileWithDefaultValues({ email: '+2339954853' })
		).rejects.toThrowErrorMatchingSnapshot();
	});
});
