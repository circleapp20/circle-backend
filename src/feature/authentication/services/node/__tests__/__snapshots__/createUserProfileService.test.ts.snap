// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createUserProfileWithDefaultValues should throw an error if email already exists 1`] = `"User already exists"`;

exports[`#createUserProfileWithDefaultValues should throw bad request error if phone number already exists 1`] = `"User already exists"`;
